Trial tracemalloc as memory tracker (replacement v2 PR)

pp-mo opened the pull request:
# ***WIP DO NOT MERGE THIS, EVER***

Some modified code for memory benchmarks, evaluating "tracemalloc" against our existing Linux process-RSS measuring technique.


### Additional context : sample output from a desktop run
```
$ asv run -b Memcheck* --quick -e -E existing:/tmp/persistent/newconda-envs/iris3/bin/python 
Couldn't load asv.plugins._mamba_helpers because
No module named 'libmambapy'
· Ignoring ASV setting(s): `python`. Benchmark environment management is delegated to third party script(s).
· Discovering benchmarks
· Running 4 total benchmarks (1 commits * 1 environments * 4 benchmarks)
[ 0.00%] ·· Benchmarking existing-py_tmp_persistent_newconda-envs_iris3_bin_python
[ 0.00%] ··· Importing benchmark suite produced output:
[ 0.00%] ···· /net/home/h05/itpp/git/iris/iris_main/benchmarks/benchmarks/generate_data/__init__.py:52: UserWarning: No BENCHMARK_DATA env var, defaulting to /net/home/h05/itpp/git/iris/iris_main/benchmarks/.data. Note that some benchmark files are GB in size.
              warn(message)

[12.50%] ··· memtrace_evaluation.MemcheckBlocksAndWorkers.time_calc                                                                                                                                 ok
[12.50%] ··· ========= ========== ========= =========
             --                   nworkers           
             --------- ------------------------------
              nblocks      1          4         9    
             ========= ========== ========= =========
                 1      568±0ms    543±0ms   514±0ms 
                 4      144±0ms    181±0ms   151±0ms 
                 9      55.9±0ms   116±0ms   109±0ms 
             ========= ========== ========= =========

[25.00%] ··· memtrace_evaluation.MemcheckBlocksAndWorkers.track_addmem_calc                                                                                                                         ok
[25.00%] ··· ========= ======= ======= =======
             --                nworkers       
             --------- -----------------------
              nblocks     1       4       9   
             ========= ======= ======= =======
                 1      305.3   305.3   305.3 
                 4       19.2    76.4    76.4 
                 9       4.0     15.2    33.9 
             ========= ======= ======= =======

[37.50%] ··· memtrace_evaluation.MemcheckRunstyles.time_calc                                                                                                                                        ok
[37.50%] ··· ============= ================= ================= =================== ===================
             --                                          runtype / ysize                              
             ------------- ---------------------------------------------------------------------------
                measure     threads / 10000   threads / 40000   processes / 10000   processes / 40000 
             ============= ================= ================= =================== ===================
              tracemalloc       91.2±0ms          228±0ms            40.2±0ms            82.4±0ms     
                  rss           87.0±0ms          222±0ms            45.5±0ms            81.5±0ms     
             ============= ================= ================= =================== ===================

[50.00%] ··· memtrace_evaluation.MemcheckRunstyles.track_addmem_calc                                                                                                                                ok
[50.00%] ··· ============= ================= ================= =================== ===================
             --                                          runtype / ysize                              
             ------------- ---------------------------------------------------------------------------
                measure     threads / 10000   threads / 40000   processes / 10000   processes / 40000 
             ============= ================= ================= =================== ===================
              tracemalloc         17.0              67.9               0.2                 0.2        
                  rss             9.1               34.4               0.9                 0.9        
             ============= ================= ================= =================== ===================
```

pp-mo commented:
## What I think we can deduce from the benchmark results
  1. neither "rss" or "tracemalloc" tracks memory in subprocesses
     * compare `MemcheckRunstyles.track_addmem_calc` , any "threads" with "processes"
  2. "rss" method is not picking up on memory used in multiple threads, but "tracemalloc" is
     * compare `MemcheckRunstyles.track_addmem_calc` , "threads" results ..
     * .. these are ~2 times as big for "tracemalloc" as for "rss" (this for nblocks=6, nworkers=4)

pp-mo commented:
### Updated ideas
@trexfeathers pointed out that the above were run on a machine with only 4 CPUs (hyperthread), hence nworkers=4/9 can be expected to be questionable.
Also a bit difficult to interpret -- questioned whether the nworkers/nblocks axes were correct !

Hence changes in https://github.com/SciTools/iris/pull/5946/commits/d039ad03bc2b131a9691f4f6e7e2bdbb3b74f9cb

  * n_workers defaults  to 3
  * n_workers = [1, 2, 3, 4] in BlocksAndWorkers testing
  * MemcheckBlocksAndWorkers_Rss repeats the n/n matrix with the Rss measure

### New results
typical run ...
```
$ asv run -b Memcheck --quick -e -E existing:/tmp/persistent/newconda-envs/iris3/bin/python
---
    Couldn't load asv.plugins._mamba_helpers because
    No module named 'libmambapy'
    · Ignoring ASV setting(s): `python`. Benchmark environment management is delegated to third party script(s).
    · Discovering benchmarks
    · Running 6 total benchmarks (1 commits * 1 environments * 6 benchmarks)
    [ 0.00%] ·· Benchmarking existing-py_tmp_persistent_newconda-envs_iris3_bin_python
    [ 0.00%] ··· Importing benchmark suite produced output:
    [ 0.00%] ···· /net/home/h05/itpp/git/iris/iris_main/benchmarks/benchmarks/generate_data/__init__.py:52: UserWarning: No BENCHMARK_DATA env var, defaulting to /net/home/h05/itpp/git/iris/iris_main/benchmarks/.data. Note that some benchmark files are GB in size.
                  warn(message)

    [ 8.33%] ··· memtrace_evaluation.MemcheckBlocksAndWorkers.time_calc                                                                                                                                 ok
    [ 8.33%] ··· ========= ========== ========== ========= =========
                 --                         nworkers
                 --------- -----------------------------------------
                  nblocks      1          2          3         4
                 ========= ========== ========== ========= =========
                     1      546±0ms    508±0ms    518±0ms   517±0ms
                     4      120±0ms    156±0ms    151±0ms   139±0ms
                     9      54.6±0ms   71.2±0ms   141±0ms   122±0ms
                 ========= ========== ========== ========= =========

    [16.67%] ··· memtrace_evaluation.MemcheckBlocksAndWorkers.track_addmem_calc                                                                                                                         ok
    [16.67%] ··· ========= ======= ======= ======= =======
                 --                    nworkers
                 --------- -------------------------------
                  nblocks     1       2       3       4
                 ========= ======= ======= ======= =======
                     1      305.3   305.3   305.3   305.3
                     4       19.2    38.3    57.3    76.4
                     9       4.0     7.7     11.5    15.2
                 ========= ======= ======= ======= =======

    [25.00%] ··· memtrace_evaluation.MemcheckBlocksAndWorkers_Rss.time_calc                                                                                                                             ok
    [25.00%] ··· ========= ========== ========== ========== ==========
                 --                          nworkers
                 --------- -------------------------------------------
                  nblocks      1          2          3          4
                 ========= ========== ========== ========== ==========
                     1      501±0ms    518±0ms    570±0ms    597±0ms
                     4      121±0ms    150±0ms    132±0ms    161±0ms
                     9      52.4±0ms   67.5±0ms   64.3±0ms   64.0±0ms
                 ========= ========== ========== ========== ==========

    [33.33%] ··· memtrace_evaluation.MemcheckBlocksAndWorkers_Rss.track_addmem_calc                                                                                                                     ok
    [33.33%] ··· ========= ======= ======= ======= =======
                 --                    nworkers
                 --------- -------------------------------
                  nblocks     1       2       3       4
                 ========= ======= ======= ======= =======
                     1      305.8   305.8   305.8   305.8
                     4       19.8    38.9    20.0    20.8
                     9       4.6     8.0     11.6    15.3
                 ========= ======= ======= ======= =======

    [41.67%] ··· memtrace_evaluation.MemcheckRunstyles.time_calc                                                                                                                                        ok
    [41.67%] ··· ============= ================= ================= =================== ===================
                 --                                          runtype / ysize
                 ------------- ---------------------------------------------------------------------------
                    measure     threads / 10000   threads / 40000   processes / 10000   processes / 40000
                 ============= ================= ================= =================== ===================
                  tracemalloc       97.5±0ms          208±0ms            39.9±0ms            94.1±0ms
                      rss           49.5±0ms          203±0ms            51.5±0ms            91.6±0ms
                 ============= ================= ================= =================== ===================

    [50.00%] ··· memtrace_evaluation.MemcheckRunstyles.track_addmem_calc                                                                                                                                ok
    [50.00%] ··· ============= ================= ================= =================== ===================
                 --                                          runtype / ysize
                 ------------- ---------------------------------------------------------------------------
                    measure     threads / 10000   threads / 40000   processes / 10000   processes / 40000
                 ============= ================= ================= =================== ===================
                  tracemalloc         12.8              51.0               0.2                 0.2
                      rss             13.2              51.1               0.8                 0.8
                 ============= ================= ================= =================== ===================

```

### Some interpretation
#### RunStyle checks are now more understandable.
```
MemcheckRunstyles.time_calc (run **WITHOUT** --quick)
      --                                          runtype / ysize
      ------------- ---------------------------------------------------------------------------
         measure     threads / 10000   threads / 40000   processes / 10000   processes / 40000
      ============= ================= ================= =================== ===================
       tracemalloc       58.2±8ms          179±5ms            44.3±1ms            86.4±3ms
           rss           71.1±6ms          171±8ms            43.7±3ms            91.2±3ms
      ============= ================= ================= =================== ===================
```
**BUT N.B.** results are *still* not very stable from time to time
```
      ============= ================= ================= =================== ===================
       tracemalloc       62.9±5ms          172±10ms           41.8±4ms            89.1±6ms
           rss           62.3±7ms          173±9ms            49.1±4ms           91.3±0.9ms
      ============= ================= ================= =================== ===================
       tracemalloc       58.2±3ms          152±20ms           44.9±5ms            91.8±5ms
           rss          58.7±10ms          170±7ms            46.1±2ms            91.6±7ms
      ============= ================= ================= =================== ===================
```
((but generally, ARE about the same for both methods, as we would expect))
**conclude** : 
  * ~same for methods
  *  processes slightly quicker than threads generally (??odd?)

#### MemcheckBlocksAndWorkers
**General picture**
```
memtrace_evaluation.MemcheckBlocksAndWorkers.time_calc
     --                         nworkers
     --------- -----------------------------------------
      nblocks      1          2          3         4
     ========= ========== ========== ========= =========
         1      546±0ms    508±0ms    518±0ms   517±0ms
         4      120±0ms    156±0ms    151±0ms   139±0ms
         9      54.6±0ms   71.2±0ms   141±0ms   122±0ms
     ========= ========== ========== ========= =========

memtrace_evaluation.MemcheckBlocksAndWorkers.track_addmem_calc
     --                    nworkers
     --------- -------------------------------
      nblocks     1       2       3       4
     ========= ======= ======= ======= =======
         1      305.3   305.3   305.3   305.3
         4       19.2    38.3    57.3    76.4
         9       4.0     7.7     11.5    15.2
     ========= ======= ======= ======= =======
```
**now makes more sense** : can see memory reducing with nblocks, and increasing with nworkers

**compare methods**
```
memtrace_evaluation.MemcheckBlocksAndWorkers_Rss.track_addmem_calc
     --                    nworkers
     --------- -------------------------------
      nblocks     1       2       3       4
     ========= ======= ======= ======= =======
         1      305.8   305.8   305.8   305.8
         4       19.8    38.9    20.0    20.8
         9       4.6     8.0     11.6    15.3
     ========= ======= ======= ======= =======

```
**conclude**
  *  results are *mostly* similar, 
  * ... except RSS tends to under-read ..
  *  .. and ***EXCEPT for nblocks=4***
    * TYPICAL (see e.g. below)

|  /nworkers | 1 | 2 | 3 | 4 |
| --- | --- | --- | --- | --- |
| tracemalloc |  19.2  |  38.3 |   57.3 |   76.4 |
| rss               | 19.8   | 38.9  |  20.0  |  20.8 |

  * POSSIBLY
    * RSS shows re-use of blocks within the operation ?
    * .. or RSS "misses" some pages ?
  * generally, it is **NOT** like RSS is only tracking 1 worker
    * since, with many blocks, mem-usage generally grows with nworkers, just as for tracemalloc

pp-mo commented:
## Outcome
  * methods are ~equivalent in scope, 
    * both can measure across threads  
    * neither can measure across processes
  * but tracemalloc is probably more reliable
    * rss results are les repeatable
    * rss seems to under-read in some cases

stephenworsley commented on benchmarks/benchmarks/memtrace_evaluation/__init__.py:
```diff
@@ -0,0 +1,122 @@
+        nblocks = params["nblocks"]
+        nworkers = params["nworkers"]
+
+        nyfull = ysize // nblocks
```
It looks like you're applying this division twice, I think that's why adding blocks has such an exagerated effect on performance.

pp-mo replied:
Oh, yes I see !
Hopefully that may explain the peculiar behaviour of the timings too.

I will fix this and re-investigate ...

pp-mo replied:
https://github.com/SciTools/iris/pull/5946/commits/84ff4ad5d6cf9c8b95a204dd563764c36ec27f69 fixes, I think

pp-mo commented on benchmarks/benchmarks/memtrace_evaluation/memory_exercising_task.py:
```diff
@@ -0,0 +1,94 @@
+                for i_task in range(self.n_blocks)
+            ],
+        )
+        combined = np.stack(list(partial_results))
```
No I _think_ this is correct according to my original plan ...
Of course it's all a bit arbitrary, but the basic idea was to create a "blocks" dimension and average over that.

It should all be explained here : https://github.com/pp-mo/iris/blob/d039ad03bc2b131a9691f4f6e7e2bdbb3b74f9cb/benchmarks/benchmarks/memtrace_evaluation/memory_exercising_task.py#L11-L18

Do you think what it is actually doing differs from that description ?

pp-mo commented:
Conclusions:
Since we fixed the task-size scaling (and added a process-based BlocksAndWorkers), the results are now roughly :
  * Runstyle checks show that old + new memory schemes delivery roughly the same numbers
    * both methods can track memory in thread workers, but not subprocesses
    * (we find that tracemalloc delivers more stable answers -- but this was not re-retested)
  * threaded workers : space required reduces with nblocks + increases with nworkers, as you would expect
  * threaded workers : time taken is ~constant :  this seems to indicate that due to Python GIL requirements, they don't actually run in parallel
  * parallel workers : time taken scales as expected (they do run in parallel, and task division gives speedups)

pp-mo commented:
Note: the results reported in the whole-benchmark runs behave a little differently.
  - the parameter tables are laid out differently in some cases, which is odd but maybe due to some additional config in the GHA action ?
  - the memory scaling for threaded workers is different -- I suspect the number of available CPUs is different

pp-mo commented:
Added 
[Investigate repeated ops with simulated memory 'leak'.](https://github.com/SciTools/iris/pull/5946/commits/8f7166e7bfdb5c0efbef1e2b902ac1bbfc9fa2b5)

Just for reference.
A sample output ...
```
$ python bm_runner.py custom run -- -b "MemcheckTaskRepeats" -e --quick -E existing:/tmp/persistent/newconda-envs/iris3/bin/python
 . . .
    BM_RUNNER DEBUG: asv run -b MemcheckTaskRepeats -e --quick -E existing:/tmp/persistent/newconda-envs/iris3/bin/python
 . . .
· Ignoring ASV setting(s): `python`. Benchmark environment management is delegated to third party script(s).
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[ 0.00%] ·· Benchmarking existing-py_tmp_persistent_newconda-envs_iris3_bin_python
[ 0.00%] ··· Importing benchmark suite produced output:
[ 0.00%] ···· /net/home/h05/itpp/git/iris/iris_main/benchmarks/benchmarks/generate_data/__init__.py:52: UserWarning: No BENCHMARK_DATA env var, defaulting to /net/home/h05/itpp/git/iris/iris_main/benchmarks/.data. Note that some benchmark files are GB in size.
              warn(message)

[100.00%] ··· memtrace_evaluation.MemcheckTaskRepeats.track_mem                                                                                                      ok
[100.00%] ··· ======= =======
               nreps         
              ------- -------
                 1      76.8 
                 2      77.1 
                 3     102.9 
                 4     103.3 
              ======= =======
```

### Some combined results
```
                   nreps
                  ------- -------
                     1      76.8,  76.8,  76.9,  76.8,  76.8,  76.8,
                     2      77.1,  77.0,  77.1,  77.0,  77.0,  77.1,
                     3     102.8, 103.0, 103.0, 103.0, 103.0, 102.9,
                     4     103.3, 103.2, 103.5, 103.4, 103.2, 103.3,
                  ======= =======
(( incremental differences, from above numbers ))
                     1-2     0.3   0.2   0.2   0.2   0.3
                     2-3    25.7  26.0  25.9  26.0  25.8
                     3-4     0.5   0.2   0.5   0.4   0.4
                  ======= =======
```

### Conclusions
The allocation pattern is a trifle hard to explain.
Sizes **don't** relate very directly to the actual allocations
This does bring into question how stable results are likely to be.

```
